src/firejail/main.c

```diff
@@ -1286,8 +1287,15 @@ int main(int argc, char **argv, char **envp) {
 		// filtering
 		//*************************************
 #ifdef HAVE_APPARMOR
-		else if (strcmp(argv[i], "--apparmor") == 0)
+		else if (strcmp(argv[i], "--apparmor") == 0) {
 			arg_apparmor = 1;
+			strncpy(apparmor_profile,"firejail-default",29);
+		}
+		else if (strncmp(argv[i], "--apparmor=",11) == 0) {
+			arg_apparmor = 1;
+			strncpy(apparmor_profile,argv[i]+11,29);
+			apparmor_profile[strlen(argv[i])>30?30:strlen(argv[i])]='\0';
```

Comment:
`apparmor_profile[30]` will not work, as it would write the 31st byte of `apparmor_profile` (as counting starts at 0).  
it would be sufficient to memset `apparmor_profile` to 0 before copying the string. because after copying either a null-terminated string got copied, or 29 bytes got copied not null-terminated, but in this case there would already be a 0 at the end of the buffer.

Reply:
The proposed code is also hard to read.  According to `cloc`, firejail has over
30k CLOC, so there is no point in code golfing.

See how `--hosts-file=filename` and `--netfilter=filename` are parsed in
src/firejail/main.c and their respective `fs_check_hosts_file` and
`check_netfilter_file` functions and try something similar (for example, in a
`fs_check_apparmor_profile` function).

Reply:
Sorry, I've written the wrong number to the 1297-th line. Thank you, will fix soon. I'm thinking of using **calloc** on apparmor_profile string.

Reply:
@ChrysoliteAzalea [on Jul 26](https://github.com/netblue30/firejail/pull/5274#discussion_r930321079):

> Sorry, I've written the wrong number to the 1297-th line. Thank you, will fix
> soon. I'm thinking of using **calloc** on apparmor_profile string.

strdup is also used when parsing `--hosts-file=` and `--netfilter=`, so for
consistency I think it would be better to keep it.  strdup is commonly used in
the project, while calloc is mostly only used when allocating string
arrays/larger chunks of memory:

```console
$ git grep strdup | wc -l
205
$ git grep calloc | wc -l
20
$ git grep -I calloc
src/firejail/env.c:     Env *env = calloc(1, sizeof(Env));
src/firejail/env.c:             errExit("calloc");
src/firejail/fs.c:      char **noblacklist = calloc(noblacklist_m, sizeof(*noblacklist));
src/firejail/fs.c:                              enames = calloc(npaths, sizeof(char *));
src/firejail/fs.c:                                      errExit("calloc");
src/firejail/fs.c:                              enames = calloc(2, sizeof(char *));
src/firejail/fs.c:                                      errExit("calloc");
src/firejail/fs_whitelist.c:    char **nowhitelist = calloc(nowhitelist_m, sizeof(*nowhitelist));
src/firejail/fs_whitelist.c:            errExit("calloc");
src/firejail/fs_whitelist.c:    TopDir *topdirs = calloc(TOP_MAX, sizeof(*topdirs));
src/firejail/fs_whitelist.c:            errExit("calloc");
src/firejail/fs_whitelist.c:            entry->wparam = calloc(1, sizeof(struct wparam_t));
src/firejail/fs_whitelist.c:                    errExit("calloc");
src/firejail/output.c:  char **args = calloc(argc + 1, sizeof(char *));
src/firejail/output.c:          errExit("calloc");
src/firejail/paths.c:   paths = calloc(path_cnt, sizeof(char *));
src/firejail/paths.c:           errExit("calloc");
src/firejail/sbox.c:    char **arg = calloc(num + 1, sizeof(char *));
src/firejail/sbox.c:            errExit("calloc");
```